### PR TITLE
Refactor lib to use plonk 

### LINKED ISSUE #5: Port Poseidon to Plonk
We no longer require curve25519 nor bulletproofs. 

Instead we will use Bls12_381 and plonk to work with gadgets and ZK.
So, we need to update the dependencies and refactor the code where needed.

### PR DESCRIPTION
**Pull request by CPerezz**
## Adds
- travis basic configfile
- Optimization on the Sponge hash fn to skip useless addition if the width of the messages is the appropiate one to do it. (WIDTH < HADES_WIDTH -1).
- CHANGELOG & CONTRIBUTING files.
## Changes
- Sponge hash refactor with gadget and scalar versions.

## Removes
- Old tests which no longer apply.

Closes #5

I bumped to 0.2, so if this gets merged I'll tag it but probably not release it since it's not finnished. @ZER0 let me know WDYT.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Contributing to plonk
```
Comment (ZER0) [inline CONTRIBUTING.md]: I guess it should be "Poseidon"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+to the issues we belive are easy for newcomers.
+
+If you need any help regarding an issue feel free to ask the authors in the `Draft pull request`
+or the issue (or by email if you prefer it by any reason.)
+
+# Pull Requests
+
+- Keep your pull requests as `Draft` until they are ready to be reviewed.
+
+- Be descriptive in the titles of your PRs and also provide good descriptions about
```
Comment (ZER0) [inline CONTRIBUTING.md:31]: I would also add somewhere a reference to [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)

File: Cargo.toml [hunk 3]
```diff
@@ -1,21 +1,17 @@
 [package]
 name = "poseidon252"
-version = "0.1.0"
+version = "0.2.0"
 authors = [
   "zer0 <matteo@dusk.network>"
```
Comment (ZER0) [inline Cargo.toml]: Add yourself to the list of authors, as well as Victor.

File: Cargo.toml [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
+dusk-bls12_381 = "0.1"
 merlin = "2"
 lazy_static = "1.3.0"
-hades252 = { git = "https://github.com/dusk-network/hades252" }
-
-[dependencies.bulletproofs]
-git = "https://github.com/dusk-network/bulletproofs"
-branch = "develop"
-features = ["yoloproofs"]
+hades252 = { git = "https://github.com/dusk-network/hades252", branch = "master" }
```
Comment (ZER0) [inline Cargo.toml]: Which version of Hades are we targeting?

File: src/sponge/pad.rs [hunk 5]
```diff
@@ -1,16 +1,17 @@
-use curve25519_dalek::scalar::Scalar;
+//! Padding support for sponge hash
+//!
 
-pub(crate) fn pad<T>(messages: &[T], width: usize) -> Vec<T>
+pub(crate) fn pad<T>(messages: &[T], width: usize, pad_value: T, eom: T) -> Vec<T>
```
Comment (ZER0) [inline src/sponge/pad.rs]: for consistency maybe we could call it `eom_value` at this point?

File: src/sponge/pad.rs [hunk 6]
```diff
@@ -1,16 +1,17 @@
-use curve25519_dalek::scalar::Scalar;
+//! Padding support for sponge hash
+//!
```
Comment (ZER0) [inline src/sponge/pad.rs:3]: Maybe we could be more descriptive about what it does, at least how the `pad_value` and `eom` are used.

File: src/sponge/pad.rs [hunk 7]
```diff
@@ -40,10 +42,13 @@ mod tests {
         let three = Scalar::from(3u64);
         let four = Scalar::from(4u64);
 
-        assert_eq!(&pad(&[two], 3), &[zero, two, one]);
-        assert_eq!(&pad(&[two, three], 3), &[zero, two, three, zero, one, zero]);
+        assert_eq!(&pad(&[two], 3, zero, one), &[zero, two, one]);
```
Comment (ZER0) [inline src/sponge/pad.rs]: I would also rename in the tests `zero` and `one` here as `pad_value` and `eom_value` when are used for such purpose, it would be clearer.

File: src/sponge/pad.rs [hunk 8]
```diff
@@ -56,14 +61,35 @@ mod tests {
...
+            &pad(&[two, three], 3, zero, one),
+            &[zero, two, three, zero, one, zero]
+        );
+        assert_eq!(
+            &pad(&[two, three, four], 3, zero, one),
+            &[zero, two, three, zero, four, one]
+        );
+    }
+
+    /* No longer applies.
```
Comment (ZER0) [inline src/sponge/pad.rs]: Shall we remove it, then?

File: src/sponge/sponge.rs [hunk 9]
```diff
@@ -0,0 +1,203 @@
+//! The `pad` module implements the Sponge's padding algorithm
```
Comment (ZER0) [inline src/sponge/sponge.rs:1]: It's a bit unclear to me this comment.

File: src/sponge/sponge.rs [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+use dusk_plonk::constraint_system::{StandardComposer, Variable};
+use hades252::strategies::*;
+use hades252::WIDTH;
+
+/// The `hash` function takes an arbitrary number of Scalars and returns the
+/// hash, using the `Hades` ScalarStragegy
+pub fn sponge_hash(messages: &[Scalar]) -> Scalar {
+    let mut strategy = ScalarStrategy::new();
+
+    let mut words = pad(messages, WIDTH, Scalar::zero(), Scalar::one());
```
Comment (ZER0) [inline src/sponge/sponge.rs]: For readability, I would do something like:
```rust
// Filling the vector returned by `pad` with zero
let pad_value = Scalar::zero();
// one will identify the end of messages
let eom_value = Scalar::one();

let mut words = pad(messages, WIDTH, pad_value, eom_value);
```

Or something like that.

File: src/sponge/sponge.rs [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+            strategy.perm(&mut inputs);
+            inputs
+        });
+    words[1]
+}
+
+/// The `hash` function takes an arbitrary number of plonk `Variable`s and returns the
+/// hash, using the `Hades` GadgetStragegy
+pub fn sponge_hash_gadget(composer: &mut StandardComposer, messages: &[Variable]) -> Variable {
+    let padder = composer.zero_var;
```
Comment (ZER0) [inline src/sponge/sponge.rs:46]: Ah ah ah, I like the fact that you ALREADY did what I suggested in the review for the gadget part. :) So, yeah, be consistent, you can add `padder` and `eom` also for the scalar's one.

File: src/sponge/sponge.rs [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+            let mut values = values.iter();
+            inputs.iter_mut().for_each(|input| {
+                *input = composer.add(
+                    (Scalar::one(), *input),
+                    (Scalar::one(), *values.next().unwrap()),
+                    Scalar::zero(),
+                    Scalar::zero(),
+                )
+            });
+            let mut strategy = GadgetStrategy::new(composer);
```
Comment (ZER0) [inline src/sponge/sponge.rs:73]: At this point you know better than me: since you always pass the same `composer`, do we need to create a new strategy all the time as well? Shouldn't we create it just once after the composer?

File: src/sponge/sponge.rs [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    use dusk_plonk::commitment_scheme::kzg10::PublicParameters;
+    use dusk_plonk::fft::EvaluationDomain;
+    use merlin::Transcript;
+
+    const CAPACITY: usize = 1 << 12;
+
+    fn poseidon_sponge_params(width: usize) -> (Vec<Scalar>, Scalar) {
+        let mut input = vec![Scalar::zero(); width];
+        input.iter_mut().for_each(
+            |s| *s = Scalar::one(), /*Scalar::random(&mut rand::thread_rng())*/
```
Comment (ZER0) [inline src/sponge/sponge.rs]: What's the comment line stands for

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, I would just change few nits and I've a couple of questions, but it's good overall

File: Cargo.toml (hunk 3 above)
Comment (CPerezz) [reply Cargo.toml]: Shouldn't we wait for #11 to land to then make the bump?

File: src/sponge/pad.rs (hunk 6 above)
Comment (CPerezz) [reply src/sponge/pad.rs:3]: Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag `#[deny(missing_docs)]`.
I was leaving docs for the end once #11 gets merged.

File: src/sponge/sponge.rs (hunk 9 above)
Comment (CPerezz) [reply src/sponge/sponge.rs:1]: Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag #[deny(missing_docs)].
I was leaving docs for the end once #11 gets merged.

File: src/sponge/sponge.rs (hunk 13 above)
Comment (CPerezz) [reply src/sponge/sponge.rs]: Some testing debuggin I was doing. Mb

File: src/sponge/sponge.rs (hunk 12 above)
Comment (CPerezz) [reply src/sponge/sponge.rs:73]: Since it is mutably borrowed, I cannot create the strategy at the beggining of the function since then the `&mut Composer` would be already borrowed by the strategy and I will not be able to call it's functions or attributes.

File: Cargo.toml (hunk 4 above)
Comment (CPerezz) [reply Cargo.toml]: I'm picking "master" since there's no tag or version for the newes't implementation which is using PLONK as constraint system & proof system.

I think that you were going to add the tag once the history was re-written. But I checked and it hasn't been created.

File: src/sponge/pad.rs (hunk 8 above)
Comment (CPerezz) [reply src/sponge/pad.rs]: Yepp

File: src/sponge/sponge.rs (hunk 10 above)
Comment (CPerezz) [reply src/sponge/sponge.rs]: Yep, agree!

File: src/sponge/sponge.rs (hunk 11 above)
Comment (CPerezz) [reply src/sponge/sponge.rs:46]: Done avobe!

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+which things are being added, changed, fixed or removed also addressing the corresponding
+issue that it is addressing to. See https://chris.beams.io/posts/git-commit/ for more info.
```
Comment (ZER0) [inline CONTRIBUTING.md]: For consistency, I would actually wrote:
`See [How to Write a Git Commit Message]( https://chris.beams.io/posts/git-commit/)`

File: src/sponge/pad.rs [hunk 15]
```diff
@@ -30,73 +31,73 @@ where
...
+            &pad(&[two, three, four], 3, pad_value, eom_value),
+            &[pad_value, two, three, pad_value, four, eom_value]
         );
     }
 
     #[test]
     fn test_scalar_padding_width_4() {
-        let zero = Scalar::zero();
-        let one = Scalar::one();
+        let pad_value = Scalar::zero();
```
Comment (ZER0) [inline src/sponge/pad.rs]: Since you used `padder` and `eom` as identifiers later on, I think we should be consistent and use them in the tests as well. Or viceversa. Whatever you like more.

File: Cargo.toml [hunk 16]
```diff
@@ -1,21 +1,17 @@
...
```
Comment (ZER0) [inline Cargo.toml]: As discussed, point to `0.5.0` already. Even if it fails the build on travis it will pass once we add the tag.

**Review by ZER0 — COMMENTED** [review]
LGTM, with the few nits addressed.

**Review by ZER0 — APPROVED** [review]

**Comment by CPerezz** [comment]
@ZER0 feel free to merge as soon as it passes! :)
